azkaban-common/src/main/bash/internal/startup-shared.sh

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env bash
```

Comment:
As the scripts become more complex, I'd prefer Python-based implementation over Bash.

Reply:
I think you are generally right but I slightly disagree here b/c that would assume python is available. This is generally true if you run this in general full-fledged linux environment but if you run it inside docker then it's not always there (like popular bare bone alpine image). And it doesn't quite make sense to force people to install python just for this purpose. I even think we should support bourne shell only environment for that purpose to remove the requirement for bash.

Reply:
However since bash is typically available in most environments and installing bash is easier than python (as you don't need to worry about 2/3), degenerate to sh is less of a concern here.

Reply:
I see your point.

Isn't one of the benefits of using docker to control the dependencies/environment? If the product requires Python (2 or 3), isn't it easier to make sure the requirement is met with a container?

Reply:
I think it would make sense if user jobs require python to be present. I don't think python should be installed just for the purpose of launching the nodes.

Reply:
I agree that for startup scripts, bash is sufficient. There was either a thread or Jira on the Apache Yetus mailing list that had a nice discussion of the pro's & cons of replacing bash with python. (Sorry I can't find it tonight.)

Essentially we can get 'good enough' testing by writing functions & using something like shunit2 or bats for testing. The other issue is portability. The differences between bash v3 & v4 are much less then python 2.6 or 2.7. Presuming every OS did ship with python 3.6, we would then need to build a pex file to package our python script and dependencies.

Reply:
This is a server software, not a utility library or command line tool etc that are designed to run in as many different environments as possible. 

IMHO, it is a better tradeoff to use Python than Bash for developer productivity reasons than to optimize for the ease of deployment. I understand that developers who like Bash better than Python may disagree with the benefits. It's my personal opinion. 

Bash has its place. My evaluation criteria are not the function of the code but the complexity of the code. 

There are better ways to ease deployment. Docker is one of them.

Reply:
@HappyRay I disagree. This is an OSS project, not dogfood any more. We should be thinking how users should use our software, not "no you are using it wrong" unless it's made by Apple.

I think my concerns are:
1. Switching to python introduces unnecessary dependency issues. Yes you can say "why don't you set up your environment properly?". But be aware it's tricky to have python 2 and 3 both in the system without help of other software like virtualenv or pex. What if the script assumes 3 while user scripts want 2, or the other way around? This requires either extra build system (which increases complexity in terms of your concern) or requires users to separate the environments for their scripts and azkaban's launch script properly. Both are suboptimal IMHO.
2. It does not necessarily reduce complexity. Most of the complexity in this PR comes from the fact that the logic is unnecessarily split into 2 files (discussion in #1669), in which case to keep the pid both need to be executed with `exec` and vars need to be handled carefully. We would have similar issues if we just switch to python but keep the same design. Again I don't want to side track this PR as that's not our main concern.

Reply:
> it is a better tradeoff to use Python than Bash for developer productivity reasons than to optimize for the ease of deployment

does not seem like a good argument to me considering the fact that those scripts have only been touch so many times since the beginning of the 3.x.

Reply:
@xunnanxu I am not partial to either Python or Bash, but both scripts can get overly complicated unless you minimize their role(and let the called process do most of the work). Case in point is the Hadoop launch scripts upto 2.7, after which they had to be rearchitected to ensure users understand how arguments are going to be interpreted.